pallets/loans/src/entities/loans.rs

```diff
@@ -325,19 +318,37 @@ impl<T: Config> ActiveLoan<T> {
 		Ok(())
 	}
 
-	fn ensure_can_repay(&self, amount: T::Balance) -> Result<T::Balance, DispatchError> {
-		let max_repay_amount = match &self.pricing {
-			ActivePricing::Internal(inner) => inner.calculate_debt()?,
-			ActivePricing::External(inner) => inner.calculate_debt()?,
+	fn ensure_can_repay(
+		&self,
+		mut amount: RepaidAmount<T::Balance>,
+	) -> Result<RepaidAmount<T::Balance>, DispatchError> {
+		let (interest_accrued, max_repay_principal) = match &self.pricing {
+			ActivePricing::Internal(inner) => {
+				let principal = self
```

Comment:
What if `ExternalActivePricing.outstanding_amount()` was rewritten to return separate outstanding principal and interest? Something like

```rust
#[derive(Default, Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
pub struct OutstandingAmount<Balance> {
  principal: Balance,
  interest: Balance,
}

impl OutstandingAmount<Balance> {
	pub fn total(&self) -> Result<Balance, ArithmeticError> {
		self.principal().ensure_add(self.interest())
	}
}

impl<T: Config> ExternalActivePricing<T> {
        ...

	pub fn outstanding_amount(&self) -> Result<OutstandingAmount, ArithmeticError> {
		let interest_reference_amount = self
			.info
			.notional
			.ensure_mul_int(self.outstanding_quantity)?;

		OutstandingAmount {
			principal: self.total_borrowed.ensure_sub(self.total_repaid.principal)?,
			interest: Ok(self.calculate_debt()?.ensure_sub(interest_reference_amount)?)
		}
	}

```

Reply:
Ok, I see the idea. Maybe by now, because it's only used in one place and it's not stored, could we use a tuple instead?

Reply:
> Ok, I see the idea. Maybe by now, because it's only used in one place and it's not stored nowwhere, could we use a tuple instead?

Yeah that's fine 👍 

> In the example, principal will be interest_reference_amount, right? Because total_borrowed and total_repaid are not accessible from ExternalActivePricing

Ah yeah I didn't consider that it would not be accessible from there. Never mind then